ast/property.go

```diff
@@ -78,10 +78,192 @@ func (p *PropertySubscript) rootName() string {
+				p.error("the root property must be a string subscript or a name")
+			}
+			p.next()
+			p.append(p.parseName())
+		case '[':
+			p.next()
+			p.append(p.parseSubscript())
+		default:
+			if unicode.IsSpace(rune(c)) {
+				p.error("unterminated interpolation")
+				return p.finish()
+			}
+			p.append(p.parseName())
+		}
+	}
+}
+
+// Terminates parsing. If there are no accessors (e.g. `${` or `${}`), appends an empty property name
+// accessor. Returns the rest of the string, the access, and any diagnostics.
+func (p *propertyAccessParser) finish() (string, *PropertyAccess, syntax.Diagnostics) {
+	if len(p.accessors) == 0 {
+		p.append(&PropertyName{Name: ""})
+	}
+
+	rest := p.text
+	access := &PropertyAccess{Accessors: p.accessors}
+	return rest, access, p.diags
+}
+
+// Parses a property name (e.g. `foo`).
+func (p *propertyAccessParser) parseName() *PropertyName {
+	var b strings.Builder
+	for {
+		c, ok := p.peek()
+		if !ok || p.terminatesName(c) {
+			break
+		}
+		p.next()
+		b.WriteByte(c)
+	}
+	if b.Len() == 0 {
+		p.errorf("missing property name")
+	}
+	return &PropertyName{Name: b.String()}
+}
+
+// Parses a subscript accessor (e.g. `["foo"]` or `[1]`).
+//
+// At this point we are already past the opening `[`. Consumes the terminating `]`, if any.
+func (p *propertyAccessParser) parseSubscript() *PropertySubscript {
+	c, ok := p.peek()
+	if !ok {
+		p.error("missing closing bracket in subscript")
+		return &PropertySubscript{Index: ""}
+	}
+
+	var index any
+	if c == '"' {
+		p.next()
+		index = p.parseStringSubscript()
+	} else {
+		index = p.parseIndexSubscript()
+	}
+
+	c, ok = p.peek()
+	if !ok || c != ']' {
+		p.error("missing closing bracket in subscript")
+	} else {
+		p.next()
+	}
+	return &PropertySubscript{Index: index}
+}
+
+// Parses a string subscript.
+//
+// At this point we are already past the opening `["`. Ends on EOF or an unescaped `"`. Consumes
+// the terminating `"` if any.
+func (p *propertyAccessParser) parseStringSubscript() string {
+	var propertyKey strings.Builder
+	for {
+		c, ok := p.peek()
+		if !ok {
+			p.error("missing closing quote in subscript")
+			return propertyKey.String()
+		}
+		p.next()
+
+		if c == '"' {
+			if propertyKey.Len() == 0 {
+				p.error("property key must not be empty")
+			}
+			return propertyKey.String()
+		}
+
+		if c == '\\' {
+			if n, ok := p.peek(); ok && n == '"' {
+				p.next()
+				c = n
+			}
+		}
+		propertyKey.WriteByte(c)
+	}
+}
+
+// Parses an index subscript.
+//
+// At this point we are already past the opening `[`. Ends on EOF, `]`, or a name terminator.
+// Does not consume the terminator.
+func (p *propertyAccessParser) parseIndexSubscript() any {
+	var index strings.Builder
+	for {
+		c, ok := p.peek()
+		if !ok || c == ']' || p.terminatesName(c) {
+			break
+		}
+
+		p.next()
+		index.WriteByte(c)
+	}
+
+	indexStr := index.String()
+	num, err := strconv.ParseInt(indexStr, 10, 0)
+	if err != nil {
+		p.error("invalid list index")
```

Comment:
Maybe use `p.errorf` and include the parse error?

Reply:
Deliberately didn't change these errors with this PR to simplify the diff. Will do so as a followup.